pkg/testsim/runtime.go

```diff
@@ -0,0 +1,235 @@
+	return min + time.Duration(r.Int63n(int64(max-min+1)))
+}
+
+// Runtime provides a mechanism to run a simulation with logical time.
+type Runtime struct {
+	*rand.Rand
+
+	activeProcLock sync.Mutex
+	nrActiveProcs  uint64
+	resumeCond     sync.Cond
+
+	now       int64
+	queue     actionQueue
+	queueLock sync.Mutex
+
+	stopChan chan struct{}
+}
+
+// NewRuntime creates a new simulation runtime given a source of
+// pseudo-random numbers.
+func NewRuntime(rnd *rand.Rand) *Runtime {
+	r := &Runtime{
+		Rand:     rnd,
+		stopChan: make(chan struct{}),
+	}
+	r.resumeCond = *sync.NewCond(&r.activeProcLock)
+
+	return r
+}
+
+// Spawn creates a new active process.
+func (r *Runtime) Spawn() *Process {
+	return newProcess(r)
+}
+
+// Run executes the simulation until there is no more active process
+// or scheduled action.
+func (r *Runtime) Run() {
+	r.waitQuiescence()
+	r.queueLock.Lock()
+	for r.queue.Len() > 0 {
+		r.doStepLocked()
+	}
+	r.queueLock.Unlock()
+}
+
+// RunFor works the same as Run, but stops execution after expiration
+// of the duration d of simulated time.
+func (r *Runtime) RunFor(d time.Duration) {
+	r.waitQuiescence()
+	t := r.timeAfter(d)
+	r.queueLock.Lock()
+	for r.queue.Len() > 0 && r.queue.top().deadline <= t {
+		r.doStepLocked()
+	}
+	r.queueLock.Unlock()
+}
+
+// Step executes the next scheduled action in the simulation and waits
+// until there is no more active process. It returns true if there was
+// an action to execute.
+func (r *Runtime) Step() (ok bool) {
+	r.waitQuiescence()
+	r.queueLock.Lock()
+	if r.queue.Len() > 0 {
+		r.doStepLocked()
+		ok = true
+	}
+	r.queueLock.Unlock()
+	return
+}
+
+// Stop kills all processes and terminates the simulation.
+func (r *Runtime) Stop() {
+	close(r.stopChan)
+}
+
+// timeAfter calculates the simulated time after the duration d.
+func (r *Runtime) timeAfter(d time.Duration) int64 {
+	t := r.now + int64(d)
+	if t < 0 {
+		panic("Time overflow")
+	}
+	return t
+}
+
+// scheduleAfter arranges the function f to be called after expiration
+// of the duration d of simulated time.
+func (r *Runtime) scheduleAfter(d time.Duration, f func()) *action {
+	return r.schedule(r.timeAfter(d), f)
+}
+
+// schedule arranges the function f to be called at simulated time t.
+func (r *Runtime) schedule(t int64, f func()) *action {
+	e := &action{
+		deadline: t,
+		fn:       f,
+	}
+	r.queueLock.Lock()
+	heap.Push(&r.queue, e)
+	r.queueLock.Unlock()
+	return e
+}
+
+// unschedule cancels the supplied scheduled action.
+func (r *Runtime) unschedule(e *action) {
+	r.queueLock.Lock()
+	if e.index >= 0 {
+		heap.Remove(&r.queue, e.index)
+		e.index = -1
+	}
+	r.queueLock.Unlock()
+}
+
+// doStepLocked executes the next scheduled action and then waits
+// until there is no more active process.
+func (r *Runtime) doStepLocked() {
+	e := heap.Pop(&r.queue).(*action)
+	r.now = e.deadline
+	r.queueLock.Unlock()
+	e.fn()
+	r.waitQuiescence()
```

Comment:
Exactly, we need to wait for quiescence to complete the step because the invoked action can spawn new active processes.

Reply:
Right, got it. Would it make sense to not activate the processes directly, but schedule them at randomized instants in the future (virtual time - wise)?
E.g., if a process writes to several channels, it seems like the corresponding processes would always be activated in the same order (order of writing to the channels). If they were not directly activated, but only scheduled for future execution, could that capture more of the native non-determinism in Go's native scheduling?
This is just a thought about a potential extension. It definitely works the way it is and unless the extension makes sense and its implementation is trivial, let's leave it for later.

Pls. mark this comment as resolved when you read this.

Reply:
I was thinking about that, but the difficulty is that the simulation still works around the Go runtime. With the current semantics, we cannot block in `Spawn` or `Fork`. Maybe we could redefine them such that they would work more like Go's `go` statement, i.e. receive a function as a parameter and run that function in the new process' context. However, I'm not sure how convenient that would be to use.

So far we explicitly `Yield` to avoid concurrency and `Delay` for random duration to simulate non-determinism of concurrent goroutines.